File: src/go/problems/dest_city_test.go
// https://leetcode.com/problems/destination-city

package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type DestCity struct{}

func TestDestCity(t *testing.T) {
	gen := core.TestSuite[DestCity]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param([][]string{{"London", "New York"}, {"New York", "Lima"}, {"Lima", "Sao Paulo"}}).Result("Sao Paulo")
	}).Add(func(tc *core.TestCase) {
		tc.Param([][]string{{"B", "C"}, {"D", "B"}, {"C", "A"}}).Result("A")
	}).Add(func(tc *core.TestCase) {
		tc.Param([][]string{{"A", "Z"}}).Result("Z")
	}).Run(t)
}

func (DestCity) Solution(paths [][]string) string {
	from := map[string]bool{}
	for i := 0; i < len(paths); i++ {
		from[paths[i][0]] = true
	}

	for _, path := range paths {
		if !from[path[1]] {
			return path[1]
		}
	}

	return ""
}

File: src/challenges/python/problems/dest_city_test.py
# https://leetcode.com/problems/destination-city

from core.problem_base import *

class DestCity(ProblemBase):
    def Solution(self, paths: list[list[str]]) -> str:
        map = {}
        for path in paths:
            map[path[0]] = map.get(path[0], 0) + 1
            map[path[1]] = map.get(path[1], 0)

        for k, v in map.items():
            if v == 0:
                return k

        return ""

if __name__ == '__main__':
    TestGen(DestCity) \
        .Add(lambda tc: tc.Param([["London","New York"],["New York","Lima"],["Lima","Sao Paulo"]]).Result("Sao Paulo")) \
        .Add(lambda tc: tc.Param([["B","C"],["D","B"],["C","A"]]).Result("A")) \
        .Add(lambda tc: tc.Param([["A","Z"]]).Result("Z")) \
        .Run()
